hal/src/electron/modem/mdm_hal.cpp

```diff
@@ -375,6 +375,15 @@ int MDMParser::waitFinalResp(_CALLBACKPTR cb /* = NULL*/,
     return WAIT;
 }
 
+int MDMParser::command(const char* cmd, _CALLBACKPTR cb, void* param, system_tick_t timeout)
+{
+    LOCK();
+    send(cmd, strlen(cmd));
```

Comment:
This should use `MDMParser::sendFormated` to ensure it is cancellable with `if (_cancel_all_operations) return 0;`

Reply:
hm, but why is there a semantic difference between `send()` and `sendFormatted()` in regards to the `_cancel_all_operations` flag? Is it an oversight or a designed behavior?

Reply:
all send operations in mdm_hal start with sendFormatted(), so that's where I targeted the cancel operation, instead of down the line in the lower level send() function.

Reply:
ok, so every logical modem operation starts with a call to `sendFormatted()` and `send()` is mostly used to send remaining data, such as the contents of a datagram buffer. If so, could we reduce the size of [the buffer](https://github.com/spark/firmware/blob/fix/issue_1298/hal/src/electron/modem/mdm_hal.cpp#L218) used by `sendFormatted()`? Currently it allocates 1K on stack just to format an AT command and I couldn't find any place in the code where such a large buffer might be needed. Most of the commands sent via `sendFormatted()` don't even use any formatting at all and are as short as e.g. `sendFormated("AT+CGSN\r\n")`.

Reply:
I suggest we use a similar approach as in `Print::printf()`, i.e. allocate a small buffer initially and then use a variable length array if it appears the formatted data doesn't fit into the original buffer: https://github.com/spark/firmware/blob/develop/wiring/src/spark_wiring_print.cpp#L271

Reply:
That method looks fine, and we have another place in cellular_command with `char buf[256];` that could use a similar treatment.

Reply:
I'll add an overloaded `sendFormatted()` taking `va_list` as an argument in order to not duplicate the code.